tomviz/PipelineProxy.cxx

```diff
@@ -0,0 +1,791 @@
+  }
+}
+
+void PipelineProxy::removeModule(const std::string& modulePath,
+                                 const std::string& dataSourceId,
+                                 const std::string& moduleId)
+{
+  auto p = QString::fromStdString(modulePath);
+  auto i = QString::fromStdString(dataSourceId);
+  auto m = QString::fromStdString(moduleId);
+
+  // Are we removing them all?
+  if (m.isEmpty()) {
+    auto ds = findDataSource(p, i);
+    if (ds == nullptr) {
+      qCritical() << "Failed to find data source.";
+      return;
+    }
+    ModuleManager::instance().removeAllModules(ds);
+  } else {
+    auto module = findModule(p, m);
+    if (module == nullptr) {
+      qCritical() << "Failed to find module.";
+      return;
+    }
+    ModuleManager::instance().removeModule(module);
+  }
+}
+
+void PipelineProxy::removeDataSource(const std::string& dataSourcePath,
+                                     const std::string& dataSourceId)
+{
+  auto p = QString::fromStdString(dataSourcePath);
+  auto i = QString::fromStdString(dataSourceId);
+
+  // Are we removing them all?
+  if (i.isEmpty()) {
+    ModuleManager::instance().removeAllDataSources();
+  } else {
+    auto ds = findDataSource(p, i);
+    if (ds == nullptr) {
+      qCritical() << "Failed to find data source.";
+      return;
+    }
+    ModuleManager::instance().removeDataSource(ds);
+  }
+}
+
+void PipelineProxy::enableSyncToPython()
+{
+  m_syncToPython = true;
+}
+
+void PipelineProxy::disableSyncToPython()
+{
+  m_syncToPython = false;
+}
+
+void PipelineProxy::pausePipeline(const std::string& dataSourcePath)
+{
+  auto p = QString::fromStdString(dataSourcePath);
+  auto ds = findDataSource(p);
+  if (ds == nullptr) {
+    qCritical() << "Failed to find data source.";
+    return;
+  }
+
+  ds->pipeline()->pause();
+}
+
+void PipelineProxy::resumePipeline(const std::string& dataSourcePath)
+{
+  auto p = QString::fromStdString(dataSourcePath);
+  auto ds = findDataSource(p);
+  if (ds == nullptr) {
+    qCritical() << "Failed to find data source.";
+    return;
+  }
+
+  ds->pipeline()->resume();
+}
+
+void PipelineProxy::executePipeline(const std::string& dataSourcePath)
+{
+  auto p = QString::fromStdString(dataSourcePath);
+  auto ds = findDataSource(p);
+  if (ds == nullptr) {
+    qCritical() << "Failed to find data source.";
+    return;
+  }
+
+  ds->pipeline()->execute()->deleteWhenFinished();
+}
+
+bool PipelineProxy::pipelinePaused(const std::string& dataSourcePath)
+{
+  auto p = QString::fromStdString(dataSourcePath);
+  auto ds = findDataSource(p);
+  if (ds == nullptr) {
+    qCritical() << "Failed to find data source.";
+    return false;
+  }
+
+  return ds->pipeline()->paused();
+}
+
+PipelineProxyBase* PipelineProxyFactory::create()
+{
+  return new PipelineProxy();
+}
+
+void PipelineProxyFactory::registerWithFactory()
+{
+  PythonFactory::instance().setPipelineProxyFactory(new PipelineProxyFactory);
+}
+
+} // namespace tomviz
```

Comment:
So this is essentially what `PipelineStateManager` was?

Reply:
Yes, but this can be added to the factory at runtime, and avoid needing almost all Tomviz classes. It really is for runtime initialization, and avoiding sucking all of Tomviz in for what is a very simple interface.